File: fabric/src/main/java/dev/mayaqq/estrogen/fabric/datagen/loottables/EstrogenLootTables.java
package dev.mayaqq.estrogen.fabric.datagen.loottables;

import dev.mayaqq.estrogen.registry.EstrogenBlocks;
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput;
import net.fabricmc.fabric.api.datagen.v1.provider.FabricBlockLootTableProvider;

public class EstrogenLootTables extends FabricBlockLootTableProvider {

    public EstrogenLootTables(FabricDataOutput dataOutput) {
        super(dataOutput);
    }

    @Override
    public void generate() {
        add(EstrogenBlocks.CENTRIFUGE.get(), createSingleItemTable(EstrogenBlocks.CENTRIFUGE.get()));
        add(EstrogenBlocks.COOKIE_JAR.get(), createSilkTouchOnlyTable(EstrogenBlocks.COOKIE_JAR.get()));
        add(EstrogenBlocks.DORMANT_DREAM_BLOCK.get(), createSingleItemTable(EstrogenBlocks.DORMANT_DREAM_BLOCK.get()));
        add(EstrogenBlocks.ESTROGEN_PILL_BLOCK.get(), createSingleItemTable(EstrogenBlocks.ESTROGEN_PILL_BLOCK.get()));
        add(EstrogenBlocks.MOTH_WOOL.get(), createSingleItemTable(EstrogenBlocks.MOTH_WOOL.get()));
        add(EstrogenBlocks.MOTH_SEAT.get(), createSingleItemTable(EstrogenBlocks.MOTH_SEAT.get()));
    }
}

File: src/fabric/1.20.1/data/kotlin/dev/mayaqq/estrogen/datagen/impl/loottables/EstrogenLoottables.kt
package dev.mayaqq.estrogen.datagen.impl.loottables

import dev.mayaqq.estrogen.content.EstrogenBlocks
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput
import net.fabricmc.fabric.api.datagen.v1.provider.FabricBlockLootTableProvider
import net.minecraft.world.level.block.BedBlock
import net.minecraft.world.level.block.state.properties.BedPart
import net.minecraft.world.level.storage.loot.functions.CopyNbtFunction
import net.minecraft.world.level.storage.loot.providers.nbt.ContextNbtProvider


class EstrogenLoottables(output: FabricDataOutput) : FabricBlockLootTableProvider(output) {

    override fun generate() {
        add(EstrogenBlocks.CookieJar, createSilkTouchOnlyTable(EstrogenBlocks.CookieJar))
        add(EstrogenBlocks.DreamBlock, createSingleItemTable(EstrogenBlocks.DreamBlock))
        add(EstrogenBlocks.EstrogenPillBlock, createSingleItemTable(EstrogenBlocks.EstrogenPillBlock))
        add(EstrogenBlocks.MothWool, createSingleItemTable(EstrogenBlocks.MothWool))
        add(EstrogenBlocks.QuiltedMothWool, createSingleItemTable(EstrogenBlocks.QuiltedMothWool))
        add(EstrogenBlocks.QuiltedMothCarpet, createSingleItemTable(EstrogenBlocks.QuiltedMothCarpet))
        add(EstrogenBlocks.MothCarpet, createSingleItemTable(EstrogenBlocks.MothCarpet))
        add(
            EstrogenBlocks.MothBed,
            createSinglePropConditionTable(EstrogenBlocks.MothBed, BedBlock.PART, BedPart.HEAD)
        )
        add(
            EstrogenBlocks.QuiltedMothBed,
            createSinglePropConditionTable(EstrogenBlocks.QuiltedMothBed, BedBlock.PART, BedPart.HEAD)
        )
        add(EstrogenBlocks.DreamCatcher, createSingleItemTable(EstrogenBlocks.DreamCatcher).apply(
            CopyNbtFunction.copyData(ContextNbtProvider.BLOCK_ENTITY)
                .copy(
                "colors",
                "colors"
            )
        ))
    }
}